Initial Kafka setup

Roshan-Mehta opened the pull request:
## Fixes

## Changes

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

divyank-aggarwal committed:
initial setup for kafka and basic use of kafka to send events data for processing to 0box

divyank-aggarwal committed:
corrected setup after testing

divyank-aggarwal committed:
refactoring kafka

devyetii committed:
resolve simple comments

devyetii committed:
kafka support multiple topics

devyetii committed:
publish events after processing them

devyetii committed:
publish events async

devyetii committed:
finalize event sequece number with manual testing

devyetii committed:
mod tidy

devyetii committed:
counter for round events not global

devyetii committed:
separate kafka in a separate compose file

devyetii committed:
clean kafka within clean.sh

devyetii committed:
global counter

devyetii committed:
revert conductor workflow files

devyetii committed:
revert

Roshan-Mehta committed:
Revert unwanted changes

Roshan-Mehta committed:
Revert go.work.sum

devyetii committed:
fix missing dependency

devyetii committed:
go mod tidy

devyetii committed:
Merge branch 'sprint-1.13' into kafka_stream_2

devyetii committed:
fix db migrations

Roshan-Mehta committed:
Publish kafka msg from addEvents

Roshan-Mehta committed:
Filter events based on data field

Roshan-Mehta committed:
Fix corner case in filter events

Roshan-Mehta committed:
Add round number

Roshan-Mehta committed:
Handle concurrent read and write safely

Roshan-Mehta committed:
Fix bug

Roshan-Mehta committed:
Send event instead of event slice

Roshan-Mehta committed:
Merge branch 'sprint-1.13' into kafka_stream_2

Roshan-Mehta committed:
fix bug in event

peterlimg commented on code/go/0chain.net/chaincore/chain/worker.go:
```diff
@@ -267,6 +267,10 @@ func (c *Chain) FinalizedBlockWorker(ctx context.Context, bsh BlockStateHandler)
 
+				if c.GetEventDb() != nil && c.GetEventDb().Store != nil {
+					go c.GetEventDb().PublishUnpublishedEvents()
+				}
```
The block is not finalized yet here, btw, why do we need to publish the unpublished events here, isn't all events are pushed to kafa when process the events when finalizing the block? i.e the `addEvents()` below. 
Also I see they are pushed to the same topic but the data are different, the `PublishUnpublisedEvents()` push the json of `event`, while the `addEvents` publish the `event.data` together with the round number. I guess this may cause issues when consumer use the data.

devyetii replied:
> The block is not finalized yet here

Can you please mention where the block is totally finalized. We can add like a worker that publishes the unpublished events.

> why do we need to publish the unpublished events here, isn't all events are pushed to kafa when process the events when finalizing the block? i.e the addEvents() below.

No, events are published only by calling `PublishUnpublishedEvents`, `addEvents` should just adds them to the DB. The DB here acts as a buffer for the events. This approach has 3 advantages:
- Decouple events publishing from event generation. Event publishing shouldn't make event generation fail if it fails.
- Events publishing failover. This provides retry for the events that wasn't pushed successfully. Thus, no assumption of a 24/7 available message broker.
- Faster block finalization at the sharder, since event storing (i.e. `addEvents`) is part of the block finalization, so storing the events to the db is definitely faster than pushing them to Kafka directly.

> Also I see they are pushed to the same topic but the data are different, the PublishUnpublisedEvents() push the json of event, while the addEvents publish the event.data together with the round number. I guess this may cause issues when consumer use the data.

They're the same. `addEvents` adds the event to the DB and `PublishUnpublishedEvents` loads it from there.

After reviewing the latest code changes, I'll add some context here. `PublishUnpublishedEvents` was the only point of contact with Kafka, @Roshan-Mehta then added the kafka publishing logic to the `addEvents`.
What I suggest now is to remove the publishing logic from addEvents and move `PublishUnpublishedEvents` to a worker on the sharder for complete separation and add limit on the events that we can store in the DB, with proper alerting if the events failing to be published exceeded a specific threshold.
This is just a suggestion, The final decision if for you guys at the end of the day, I just presented the whole context as it was when I was working on it.

Roshan-Mehta committed:
Remove PublishUnpublishedEvents

peterlimg committed:
Do not push messages to kafka async

This will cause events disordered

peterlimg committed:
Show panic logs in console

peterlimg committed:
Fix error on load snapshots

If the round 0 failed to be added due to postgres not get up yet when network just start, we should be able to update snapshots later

peterlimg committed:
Add kafka enabled cofnig

peterlimg committed:
Add logs for slow kafka push

goanchit commented on code/go/0chain.net/smartcontract/dbs/queueProvider/kafka.go:
```diff
@@ -0,0 +1,127 @@
+	)
+	if err != nil {
+		logging.Logger.Error("Publish: failed to write message on kafka", zap.String("topic", topic), zap.Any("message", message), zap.Error(err))
+		err := k.ReconnectWriter(topic)
```
@dabasov FYI

peterlimg committed:
Push kafka msg in Async with partition

peterlimg committed:
Add kafka auth

dabasov committed:
Merge pull request #3227 from 0chain/kafka-auth

Add kafka auth

dabasov reviewed (approved):
lgtm
